File: leetcode-cn/Python/275.h指数-ii.py
#
# @lc app=leetcode.cn id=275 lang=python3
#
# [275] H指数 II
#

# @lc code=start
from typing import List


class Solution:
    def hIndex(self, citations: List[int]) -> int:
        n = len(citations)
        for h in range(n, 0, -1):
            if citations[n - h] >= h:
                return h
        return 0


# @lc code=end

File: leetcode-cn/C++/275.h-指数-ii.cpp
/*
 * @lc app=leetcode.cn id=275 lang=cpp
 *
 * [275] H 指数 II
 */

// @lc code=start
#include <vector>
using namespace std;

class Solution
{
public:
    int hIndex(vector<int> &citations)
    {
        int n = citations.size();

        int left = 0;
        int right = n;

        while (left < right)
        {
            int mid = left + (right - left) / 2;

            if (citations[mid] >= n - mid)
            {
                right = mid;
            }
            else
            {
                left = mid + 1;
            }
        }

        return n - left;
    }
};
// @lc code=end
